plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -382,91 +382,83 @@ void SoundSourceM4A::restartDecoding(MP4SampleId sampleBlockId) {
-            sampleBlockId -= m_numberOfPrefetchSampleBlocks;
-        } else {
-            sampleBlockId = kSampleBlockIdMin;
+    if (m_curFrameIndex != readableFrames.start()) {
+        // NOTE(uklotzde): Resetting the decoder near to the beginning
+        // of the stream when seeking backwards produces invalid sample
+        // values! As a consequence the seeking test fails.
+        if ((m_curSampleBlockId != MP4_INVALID_SAMPLE_ID) &&
+                (readableFrames.start() < m_curFrameIndex) &&
+                (readableFrames.start() <= (frameIndexMin() + kNumberOfPrefetchFrames))) {
+            // Workaround: Reset the decoder when seeking near to the beginning
+            // of the stream while decoding.
+            reopenDecoder();
         }
-        restartDecoding(sampleBlockId);
-        DEBUG_ASSERT(m_curSampleBlockId == sampleBlockId);
-    }
 
-    // Decoding starts before the actual target position
-    DEBUG_ASSERT(m_curFrameIndex <= frameIndex);
+        MP4SampleId sampleBlockId = kSampleBlockIdMin
+                + (readableFrames.start() / m_framesPerSampleBlock);
+        DEBUG_ASSERT(isValidSampleBlockId(sampleBlockId));
+        if ((readableFrames.start() < m_curFrameIndex) || // seeking backwards?
+                !isValidSampleBlockId(m_curSampleBlockId) || // invalid seek position?
+                (sampleBlockId
+                        > (m_curSampleBlockId + m_numberOfPrefetchSampleBlocks))) { // jumping forward?
+            // Restart decoding one or more blocks of samples backwards
+            // from the calculated starting block to avoid audible glitches.
+            // Implementation note: The type MP4SampleId is unsigned so we
+            // need to be careful when subtracting!
+            if ((kSampleBlockIdMin + m_numberOfPrefetchSampleBlocks)
+                    < sampleBlockId) {
+                sampleBlockId -= m_numberOfPrefetchSampleBlocks;
+            } else {
+                sampleBlockId = kSampleBlockIdMin;
+            }
+            restartDecoding(sampleBlockId);
+            DEBUG_ASSERT(m_curSampleBlockId == sampleBlockId);
+        }
 
-    // Skip (= decode and discard) all samples up to the target position
-    const SINT prefetchFrameCount = frameIndex - m_curFrameIndex;
-    const SINT skipFrameCount = skipSampleFrames(prefetchFrameCount);
-    DEBUG_ASSERT(skipFrameCount <= prefetchFrameCount);
-    if (skipFrameCount < prefetchFrameCount) {
-        kLogger.warning() << "Failed to prefetch sample data while seeking"
-                << skipFrameCount << "<" << prefetchFrameCount;
+        // Decoding starts before the actual target position
+        DEBUG_ASSERT(m_curFrameIndex <= readableFrames.start());
+        const auto precedingFrames =
+                IndexRange::between(m_curFrameIndex, readableFrames.start());
+        if (!precedingFrames.empty()
+                && (precedingFrames != skipSampleFrames(precedingFrames))) {
```

Comment:
Is this a recursive call? How is guaranteed that we not reed the prefetched part over and over again. Can't we avoid this recursion?